### PR TITLE
feat(alert): added dismissOnTimeout to auto close alert

### LINKED ISSUE #207: Alert: self-close on timeout
It will be trivial to implement as soon as we've got #206

### PR DESCRIPTION
**Pull request by sendilkumarn**
fixes #207

### CODE REVIEW

File: src/alert/alert.ts [hunk 1]
```diff
@@ -85,7 +91,12 @@ export class NgbDismissibleAlert implements OnInit {
       this.closeEvent.emit($event);
       this.close();
     });
+    if (this.dismissOnTimeout) {
+      setTimeout(() => { this.close(); }, toInteger(this.dismissOnTimeout));
+    }
   }
+
+  ngOnDestroy() { clearInterval(this.dismissOnTimeout); }
```
Comment (ExFlo) [inline src/alert/alert.ts]: I think you need to store the result of the setTimeout() and use this with clearTimeout() in the ngOnDestroy to correctly remove the dismiss feature

File: src/alert/alert.ts [hunk 2]
```diff
@@ -63,6 +65,10 @@ export class NgbDismissibleAlert implements OnInit {
    * An event emitted when the close button is clicked.
    */
   @Output('close') closeEvent = new EventEmitter();
+  /**
+   * An event emitted when the close button is clicked.
```
Comment (ExFlo) [inline src/alert/alert.ts]: Change the comment to fit the input description, like "Amount of time in milliseconds before the alert is self closing" Thx

File: src/alert/alert.ts [hunk 3]
```diff
@@ -63,6 +66,10 @@ export class NgbDismissibleAlert implements OnInit {
    * An event emitted when the close button is clicked.
    */
   @Output('close') closeEvent = new EventEmitter();
+  /**
+   *  Time in milliseconds to auto close the alert
```
Comment (icfantv) [inline src/alert/alert.ts]: Time, in milliseconds, before the alert auto closes.

**Comment by sendilkumarn** [comment]
@icfantv think you can merge this

File: src/alert/alert.ts [hunk 4]
```diff
@@ -51,9 +53,10 @@ export class NgbAlert {
  * Alerts that can be dismissed without any additional code.
  */
 @Directive({selector: 'template[ngbAlert]'})
-export class NgbDismissibleAlert implements OnInit {
+export class NgbDismissibleAlert implements OnInit, OnDestroy {
   private _windowFactory: ComponentFactory<NgbAlert>;
   private _windowRef: ComponentRef<NgbAlert>;
+  private timeout: number;
```
Comment (Foxandxss) [inline src/alert/alert.ts]: as today we are prefixing private with underscore.

I am going to change that probably but as today, you need to change it.

**Comment by pkozlowski-opensource** [comment]
I would like to check a couple of things before merging this one. Will take care of the remaining review / merging.

File: src/alert/alert.ts [hunk 5]
```diff
@@ -63,6 +66,10 @@ export class NgbDismissibleAlert implements OnInit {
    * An event emitted when the close button is clicked.
    */
   @Output('close') closeEvent = new EventEmitter();
+  /**
+   *  Time, in milliseconds, before the alert auto closes.
+   */
+  @Input() dismissOnTimeout;
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: This should have type `number`

File: src/alert/alert.ts [hunk 6]
```diff
@@ -51,9 +53,10 @@ export class NgbAlert {
  * Alerts that can be dismissed without any additional code.
  */
 @Directive({selector: 'template[ngbAlert]'})
-export class NgbDismissibleAlert implements OnInit {
+export class NgbDismissibleAlert implements OnInit, OnDestroy {
   private _windowFactory: ComponentFactory<NgbAlert>;
   private _windowRef: ComponentRef<NgbAlert>;
+  private _timeout: number;
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: We shouldn't use `number` type here as what we get from `setTimeout` is a zone-specific data structure, not a number (`setTimeout` is patched by zone.js)

File: src/alert/alert.spec.ts [hunk 7]
```diff
@@ -99,6 +99,29 @@ describe('NgbDismissibleAlert', () => {
...
+         tick(1000);
+         fixture.detectChanges();
+         expect(alertEl.getAttribute('role')).toEqual('alert');
+         expect(getCloseButton(alertEl)).toBeTruthy();
+
+         tick(1200);
+         fixture.detectChanges();
+         expect(getAlertElement(fixture.nativeElement)).toBeNull();
+
+         discardPeriodicTasks();
```
Comment (pkozlowski-opensource) [inline src/alert/alert.spec.ts]: `discardPeriodicTasks` shouldn't be needed here as at this point all the periodic tasks should be gone.

File: src/alert/alert.spec.ts [hunk 8]
```diff
@@ -99,6 +99,29 @@ describe('NgbDismissibleAlert', () => {
...
+  it('should auto close on timeout specified', fakeAsync(inject([TestComponentBuilder], (tcb) => {
+       const html = '<template ngbAlert [dismissOnTimeout]="1000">Hello, {{name}}!</template>';
+       tcb.overrideTemplate(TestComponent, html).createAsync(TestComponent).then((fixture) => {
+         fixture.detectChanges();
+         const alertEl = getAlertElement(fixture.nativeElement);
+         fixture.detectChanges();
+         expect(alertEl.getAttribute('role')).toEqual('alert');
+         expect(getCloseButton(alertEl)).toBeTruthy();
+
+         tick(1000);
```
Comment (pkozlowski-opensource) [inline src/alert/alert.spec.ts]: Could we use a lower number (say `800`) to make it clear that we are testing what is happening before timeout?

Comment (sendilkumarn) [inline src/alert/alert.spec.ts]: yeah okay.

File: src/alert/alert.ts [hunk 9]
```diff
@@ -85,7 +92,12 @@ export class NgbDismissibleAlert implements OnInit {
       this.closeEvent.emit($event);
       this.close();
     });
+    if (this.dismissOnTimeout) {
+      this._timeout = setTimeout(() => { this.close(); }, toInteger(this.dismissOnTimeout));
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: I'm surprised that it works in a constructor... Did you test it in plunker to see if it behaves properly in the "real world"? So far I was living with an assumption that inputs are not defined in a constructor (change detection didn't run yet). 

Hmm, actually it _might_ work for literals but not when we use a variable. We need to do more testing / exploration here and eventually move the timeout-setting logic to `ngOnInit`

File: src/alert/alert.ts (hunk 5 above)
Comment (sendilkumarn) [inline src/alert/alert.ts]: can we just remove the number for now?

File: src/alert/alert.ts (hunk 6 above)
Comment (sendilkumarn) [inline src/alert/alert.ts]: can we just remove the type here?

Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: yes

File: src/alert/alert.ts (hunk 9 above)
Comment (sendilkumarn) [inline src/alert/alert.ts]: I am trying to do it in plunkr

Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: Here is a fixed plunk: http://plnkr.co/edit/yLx44InWGCX2jbsx5C6l?p=preview

Surprisingly it seems like it works in a constructor!

Comment (sendilkumarn) [inline src/alert/alert.ts]: I missed the precompile. Thanks @pkozlowski-opensource

**Comment by icfantv** [comment]
@sendilkumarn, @pkozlowski-opensource, i remember seeing a default value of 5000 here.  what was the rationale for removing it?

**Comment by pkozlowski-opensource** [comment]
I don't think we ever had ant default val for close on timeout.

**Comment by icfantv** [comment]
@pkozlowski-opensource ok.  i think it's fine for rc0.  i know i saw a 5000 somewhere.

**Comment by pkozlowski-opensource** [comment]
>  i know i saw a 5000 somewhere.

Maybe in a carousel?

**Comment by icfantv** [comment]
@pkozlowski-opensource:  oooooo, good call.  probably.

**Pull request closed without merging**